### PR TITLE
[DNM] Allow template body as parameter (instead of only template path)

### PR DESCRIPTION
**Pull request by freenerd**
This is an experimental PR that overloads the templatePath parameter:
  - retain the current functionality (pass in a string and load template fro disk or S3)
  - pass in a template body object

This enables two use cases:

- If a caller has already loaded the template into memory, they can reuse that, instead of loading it again. This significantly speeds up operations if templates are downloaded from remote S3 buckets or if they are js templates with expensive external calls.
- A caller can alter the template from disk/s3 in-code  (e.g. to validate resources for compliance or automatically add/remove resources or properties)

Todo:

- [x] Get an early opinion by @rclark: Is there a reason we did not do this before?
- [x] Code
  - [x] Code for create
  - [x] Tests for create
  - [x] Code for update
  - [x] Tests for update
  - [x] Do we need to change code for other operations?
  - [x] Changelog
- [x] code review @rclark 👀 
- [x] Release

/cc @arunasank @brendanmcfarland

### CODE REVIEW

File: lib/commands.js [hunk 1]
```diff
@@ -522,7 +522,18 @@ var operations = module.exports.operations = {
 
   createPreamble: function(context) {
     var preamble = d3.queue();
-    preamble.defer(template.read, context.templatePath, context.overrides.templateOptions);
+
+    if (Object.prototype.toString.call(context.template) === "[object String]") {
```
Comment (rclark) [inline lib/commands.js]: I'm curious why you went this way with the condition instead of one of these other options. Does this cover some edge case I'm overlooking?

```
if (context.template.toString() === '[object String]')
```

```
if (typeof context.template === 'string')
```

**Comment by freenerd** [comment]
I have fixed up the test suite and used this code successfully in my local environment in the last weeks. @rclark can I get a code review from you? Would love to get this out the door this week.
